File: utils/init/aws_initialize.py
#!/usr/bin/env python3
from bs4 import BeautifulSoup
import arn

from io import StringIO
import json
import multiprocessing
import neo4j
import os
import pandas as pd
import requests
import shutil
import time

from neo4j import GraphDatabase

NEO4J_BASE = "."
SERVICE_URL = 'https://awspolicygen.s3.amazonaws.com/js/policies.js'
url_base = "https://docs.aws.amazon.com/service-authorization/latest/reference"
services_json_url = f"{url_base}/toc-contents.json"
SERVICE_METADATA_FILE = "service_metadata.json"
MODULE_DIR = os.path.dirname(os.path.abspath(__file__))

def get_service_metadata():
    response = requests.get(SERVICE_URL)
    body_response = str(response.content)
    start_index = body_response.find('{')
    return_dictionary = {}

    metadata = json.loads(body_response[start_index:-1])
    for key, value in metadata['serviceMap'].items():
        key_name = key.replace(" ", "").lower().replace(")", "").replace("(", "")
        return_dictionary[key_name] = value

    return return_dictionary


def get_resource_type(resource_name, resources_data):
    for resource in resources_data:
        if resource_name == resource['Resource types']:
            service_name = resource['ARN'].split(':')[2]
            return f"{service_name}:{resource_name}"


def normalize_dictionary(data, resources_data):
    normalized_dictionary = {}
    for action in data:
        action_name = action['Actions'].replace(" [permission only]", "")
        action_dict = normalized_dictionary.get(action_name, {})
        action_dict['description'] = action['Description']
        if action['Access level']:
            action_dict['access_level'] = action['Access level']
        rts = action_dict.get("resource_types", [])
        rt_name = action.get("Resource types (*required)", None)
        if rt_name:
            rt_name = rt_name.replace("*", "")
            resource_type = get_resource_type(rt_name, resources_data)
            if resource_type:
                rts.append(resource_type)

        condition_keys = action_dict.get("condition_keys", [])
        ck_name = action.get("Condition keys", None)
        if ck_name:
            condition_keys.append(ck_name)

        action_dict['resource_types'] = rts
        normalized_dictionary[action_name] = action_dict

    return normalized_dictionary


def get_table_data(table):
    html_io = StringIO(str(table))
    df = pd.read_html(html_io)[0]
    table_json = df.to_json(orient='records')
    table_data = json.loads(table_json)
    return table_data


def html_tables_to_json(html_content):

    # Parse HTML with BeautifulSoup
    soup = BeautifulSoup(html_content, 'html.parser')
    actions_data = []
    resources_data = []
    conditions_data = []

    # Find the first table in the HTML
    tables = soup.findAll('table')

    for table in tables:
        table_name = table.th.text
        if table_name == "Actions":
            actions_data = get_table_data(table)
        elif table_name == "Resource types":
            resources_data = get_table_data(table)
        elif table_name == "Condition keys":
            conditions_data = get_table_data(table)
        else:
            print(f"[*] Unknown table type: {table_name}")

    return actions_data, resources_data, conditions_data


def extract_hrefs(data):
    href_list = data['contents'][0]['contents'][0]['contents']
    hrefs = list(map(lambda x: x['href'], href_list))
    return hrefs


def get_service_dict(ret_data, link):
    try:
        service_name = link.removesuffix('.html').removeprefix("list_")
        service_name = service_name.replace(" ", "").lower()
        response = requests.get(f"{url_base}/{link}")
        response.raise_for_status()

        html_content = response.text
        actions_json, resources_json, condition_json = html_tables_to_json(
            html_content)
        service_dict = {}
        service_dict['Actions'] = normalize_dictionary(
            actions_json,
            resources_json)
        service_dict['ResourceTypes'] = resources_json
        service_dict['ConditionKeys'] = condition_json
        ret_data[service_name] = service_dict

    except requests.RequestException as e:
        print(f"Failed to fetch content from {link}. Reason: {e}")


def get_all_services_html(data):
    all_hrefs = extract_hrefs(data)
    manager = multiprocessing.Manager()
    ret_data = manager.dict()
    jobs = []
    for link in all_hrefs:
        p = multiprocessing.Process(target=get_service_dict,
                                    args=(ret_data, link))
        jobs.append(p)
        p.start()

    for proc in jobs:
        proc.join()

    return ret_data.copy()


def get_service_prefix(service_name):
    if not os.path.exists(SERVICE_METADATA_FILE):
        print("[*] No existing service metadata file. Building...")
        return_dictionary = get_service_metadata()
        with open(SERVICE_METADATA_FILE, 'w') as f:
            f.write(json.dumps(return_dictionary, sort_keys=True, indent=4))

    with open(SERVICE_METADATA_FILE) as f:
        service_data = json.loads(f.read())
        if service_name not in service_data:
            print(f"[*] Service name: {service_name} not found")
            return None
        return service_data[service_name]["StringPrefix"]


def write_data_to_csv(actions: dict, actions_file, resources_file,
                      condition_keys_file, action_to_resources_types_rels_file):
    actions_file.write("name,access_level\n")
    resources_file.write("name,arn\n")
    action_to_resources_types_rels_file.write("source,dest\n")
    for service_name, definition in actions.items():
        service_prefix = get_service_prefix(service_name)
        lines = []
        for action_name, action_def in definition["Actions"].items():
            full_name = f"{service_prefix}:{action_name}"
            access_level = action_def['access_level']
            lines.append(f"{full_name},{access_level}\n".lower())
            for resource in action_def.get('resource_types', []):
                action_to_resources_types_rels_file.write(
                    f"{full_name},{resource}\n".lower()
                )

        actions_file.writelines(lines)

        lines = []
        resource_types = definition.get("ResourceTypes", [])
        for resource_type in resource_types:
            name = resource_type['Resource types']
            resource_arn = resource_type['ARN']
            rt_service = arn.Arn.fromstring(resource_arn).service
            full_name = f"{rt_service}:{name}"
            lines.append(f"{full_name},{resource_arn}\n".lower())
        resources_file.writelines(lines)

        lines = []
        condition_keys = definition.get("ConditionKeys", [])
        for condition_key in condition_keys:
            name = condition_key['Condition keys']
            lines.append(f"{name}\n".lower())
        condition_keys_file.writelines(lines)


def get_services_json():
    response = requests.get(services_json_url)
    return json.loads(response.text)


def ingest_csv(session, filename, datatype, fields):    
    query = (
        f'LOAD CSV FROM "file:///{filename}" AS row WITH '
    )

    row_names = ", "
    rows = []
    for i in range(len(fields)):
        rows.append(f"row[{i}] as {fields[i]}")

    query += row_names.join(rows)

    query += (
        f" MERGE (a:{datatype} {{{fields[0]}:{fields[0]}}}) "
        "ON CREATE SET "
    )

    for field in fields:
        query += f"a.{field} = {field}, "

    query += "a.layer = 0"
    session.run(query)

def ingest_relationships(session, filename, source_label, source_field,
                         rel_name, dest_label, dest_field):
    print(f"[*] Processing relationship: {filename}")
    query = (
        f'LOAD CSV FROM "file:///{filename}" AS row '
        'CALL { '
        'WITH row '
        f'MERGE (s:{source_label} {{{source_field}: row[0]}}) '
        f'ON CREATE SET s.inferred = true '
        f'MERGE (d:{dest_label} {{{dest_field}: row[1]}}) '
        f'ON CREATE SET d.inferred = true '
        f'MERGE (s) - [:{rel_name} {{layer: 0}}] -> (d) '
        '} IN TRANSACTIONS'
    )
    session.run(query)


def load_csvs_into_database(driver):


    with driver.session() as session:
        ingest_csv(session, "awsmultivaluedprefix.csv",
                   "AWSMultivalueOperator:UniqueName",
                   ["name"])
        ingest_csv(session, "awsoperators.csv", "AWSOperator:UniqueName",
                   ["name"])
        ingest_csv(session, "awsresourcetypes.csv", "AWSResourceType:UniqueName",
                   ["name", "arn"])
        ingest_csv(session, "awsglobalconditionkeys.csv", "AWSConditionKey:UniqueName",
                   ["name"])
        ingest_csv(session, "awsconditionkeys.csv", "AWSConditionKey:UniqueName",
                   ["name"])
        ingest_csv(session, "awsactions.csv", "AWSAction:UniqueName",
                   ["name", "access_level"])
        ingest_relationships(session, "actions_to_resourcetypes_rels.csv", "AWSAction:UniqueName",
                             "name", "ActsOn", "AWSResourceType:UniqueName", "name")

def create_constraint(session, constraint_name, label, property):    
    query = (
        f"CREATE CONSTRAINT {constraint_name} IF NOT EXISTS "
        f"FOR (n:{label}) REQUIRE n.{property} IS UNIQUE")

    session.run(query)

def create_relationship_constraint(session, constraint_name, rel_name,
                                   unique_property_name):
    query = (
        f"CREATE CONSTRAINT {constraint_name} IF NOT EXISTS "
        f"FOR () - [r:{rel_name}] -() REQUIRE (r.{unique_property_name}) IS UNIQUE"
    )
    session.run(query)


def create_constraints(driver):
    with driver.session() as session:
        create_constraint(session, 'awsactionconstraint',
                          'AWSAction', 'name')
        create_constraint(session, 'actionblobconstraint',
                          'AWSActionBlob', 'name')
        create_constraint(session, 'assumerolepolicyconstraint',
                          'AWSAssumeRolePolicy', 'hash')
        create_constraint(session, "conditionconstraint",
                          "AWSCondition", "hash")
        create_constraint(session, "conditionvalueconstraint",
                          "AWSConditionValue", "name")
        create_constraint(session, "groupconstraint",
                          "AWSGroup", "arn")
        create_constraint(session, "inlinepolicyconstraint",
                          "AWSInlinePolicy", "hash")
        create_constraint(session, "managedpolicyconstraint",
                          "AWSManagedPolicy", "arn")
        create_constraint(session, "policydocumentconstraint",
                          "AWSPolicyDocument", "hash")
        create_constraint(session, "policyversionconstraint",
                          "AWSPolicyVersion", "hash")
        create_constraint(session, "roleconstraint",
                          "AWSRole", "arn")
        create_constraint(session, "statementconstraint",
                          "AWSStatement", "hash")
        create_constraint(session, "userconstraint",
                          "AWSUser", "arn")
        create_constraint(session, "resourceblobconstraint",
                          'AWSResourceBlob', 'name')
        create_constraint(session, "tagconstraint",
                          'AWSTag', 'hash')

def create_indices(driver):
    with driver.session() as session:
        query = (
            'CREATE TEXT INDEX uniquehash IF NOT EXISTS '
            'FOR (n:UniqueHash) ON (n.hash)'
        )
        session.run(query)
        query = (
            'CREATE TEXT INDEX uniquearn IF NOT EXISTS '
            'FOR (n:UniqueArn) ON (n.arn)'
        )
        session.run(query)
        query = (
            'CREATE TEXT INDEX uniquename IF NOT EXISTS '
            'FOR (n:UniqueName) ON (n.name)'
        )
        session.run(query)
        query = (
            'CREATE TEXT INDEX statementeffect IF NOT EXISTS '
            'FOR (s:AWSStatement) ON (s.effect)'
        )
        session.run(query)
        query = (
            'CREATE TEXT INDEX actionname IF NOT EXISTS '
            'FOR (s:AWSAction) ON (s.name)'
        )
        session.run(query)
        query = (
            'CREATE TEXT INDEX actionblobname IF NOT EXISTS '
            'FOR (s:AWSActionBlob) ON (s.name)'
        )
        session.run(query)
        query = (
            'CREATE TEXT INDEX roleblobname IF NOT EXISTS '
            'FOR (s:AWSResourceBlob) ON (s.name)'
        )
        session.run(query)
        query = (
            'CREATE TEXT INDEX rolebname IF NOT EXISTS '
            'FOR (s:AWSResource) ON (s.name)'
        )
        session.run(query)


def aws_initialize(output_dir, schema_path=None):

    if not schema_path:
        services_json = get_services_json()
        aws_schema = get_all_services_html(services_json)
        print("[*] Writing new schema to awschema.json")
        with open('awsschema.json', 'w') as f:
            f.write(json.dumps(aws_schema, indent=4))
    else:
        try:
            with open('awsschema.json', 'r') as f:
                print("[*] Loading existing schema")
                aws_schema = json.loads(f.read())
        except FileNotFoundError:
            print("[!] Schema file not found. Rerun without input param to generate new schema file")

    if not os.path.exists(output_dir):
        os.makedirs(output_dir)

    actions_filename = os.path.join(output_dir, "awsactions.csv")
    resources_filename = os.path.join(output_dir, "awsresourcetypes.csv")
    ck_filename = os.path.join(output_dir, "awsconditionkeys.csv")
    rt_rels_filename = os.path.join(output_dir, "actions_to_resourcetypes_rels.csv")

    with open(
            actions_filename, 'w') as actions_file, open(
            resources_filename, 'w') as resources_file, open(
            ck_filename, 'w') as ck_file, open(
                rt_rels_filename, 'w') as action_to_rt_rels_file:
        write_data_to_csv(aws_schema, actions_file, resources_file,
                          ck_file, action_to_rt_rels_file)

    # These files are manually maintained
    shutil.copy(os.path.join(MODULE_DIR, "awsoperators.csv"),
                os.path.join(output_dir, "awsoperators.csv"))
    shutil.copy(os.path.join(MODULE_DIR, "awsglobalconditionkeys.csv"),
                os.path.join(output_dir, "awsglobalconditionkeys.csv"))
    shutil.copy(os.path.join(MODULE_DIR, "awsmultivaluedprefix.csv"),
                os.path.join(output_dir, "awsmultivaluedprefix.csv"))

    while (True):
        try:
            driver = GraphDatabase.driver("bolt://localhost:7687", 
                                          auth=None)
            load_csvs_into_database(driver)
            create_constraints(driver)
            create_indices(driver)
            break
        except neo4j.exceptions.ServiceUnavailable:
            print("[*] Failed to connect retrying in 5 seconds")
            time.sleep(5)


if __name__ == "__main__":
    import argparse
    argParser = argparse.ArgumentParser()
    argParser.add_argument("-o", "--output-directory",
                           help="The directory to output csv files", required=True)
    argParser.add_argument("-i", "--input-schema", help="The AWS input schema", 
                           default=None)
    args = argParser.parse_args()
    aws_initialize(args.output_directory)
    print("[*] Finished")

File: utils/init/aws_initialize.go
package main

import (
	"context"
	"encoding/csv"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"os"
	"regexp"
	"strconv"
	"strings"

	"github.com/PuerkitoBio/goquery"
	"github.com/neo4j/neo4j-go-driver/v5/neo4j"
)

// PolicyEditorConfig represents the top-level JSON structure
type PolicyEditorConfig struct {
	PolicyEditorConfig_1 []string                  `json: "app.PolicyEditorConfig"`
	ConditionOperators   []string                  `json:"conditionOperators"`
	ConditionKeys        []string                  `json:"conditionKeys"`
	ServiceMap           map[string]ServiceDetails `json:"serviceMap"`
}

// ServiceDetails represents the details of each service in the service map
type ServiceDetails struct {
	//StringPrefix string   `json:"StringPrefix"`
	Actions     []string `json:"Actions"`
	ARNFormat   string   `json:"ARNFormat"`
	ARNRegex    string   `json:"ARNRegex"`
	HasResource bool     `json:"HasResource"`
}

type Content struct {
	Title    string    `json:"title"`
	Href     string    `json:"href"`
	Contents []Content `json:"contents,omitempty"`
}

type Root struct {
	Contents []Content `json:"contents,omitempty"`
}

type Actions struct {
	Action        string
	Description   string
	Accesslevel   string
	ResourceType  string
	ConditionKeys string
}

func printSubsections(content Content) {
	for _, subContent := range content.Contents {
		fmt.Printf("Title: %s\n", subContent.Title)
		fmt.Printf("Href: %s\n", subContent.Href)
	}
}

func get_service_metadata(url string) (*PolicyEditorConfig, error) {

	response, err := http.Get(url)
	error_check(err)

	defer response.Body.Close()

	// Read the response body
	body, err := ioutil.ReadAll(response.Body)
	error_check(err)

	// Define the structure to store the JSON data
	var config PolicyEditorConfig

	// Unmarshal the JSON data into the struct
	output := regexp.MustCompile("app.PolicyEditorConfig=")
	edit := output.ReplaceAllString(string(body), "")

	err = json.Unmarshal([]byte(edit), &config)
	if err != nil {
		error_check(err)
	}

	return &config, nil
}

func get_services_json_href(url string) map[string]string {

	//web request
	response, err := http.Get(url)
	error_check(err)

	defer response.Body.Close()

	body, err := ioutil.ReadAll(response.Body)
	error_check(err)

	//parse json data
	var contents Root
	err = json.Unmarshal(body, &contents)
	if err != nil {
		fmt.Println("Error with unmarshal:", err)
	}

	//var serviceList []string

	serviceList := make(map[string]string)

	for _, data := range contents.Contents {
		//fmt.Println(data.Contents)
		for _, y := range data.Contents {
			//printSubsections(y)
			for _, subContent := range y.Contents {

				serviceList[subContent.Title] = subContent.Href
			}

		}
	}

	return serviceList

}

func get_service_dict(link string) ([]Actions, error) {
	//grab href from services json file, and pull the actions from the link

	response, err := http.Get(link)
	error_check(err)

	defer response.Body.Close()

	doc, err := goquery.NewDocumentFromReader(response.Body)
	error_check(err)

	var details []Actions
	re := regexp.MustCompile(" +")

	//table #may change with each html response so I should figure out how to add programatically
	doc.Find("table").Eq(0).Each(func(column int, tr *goquery.Selection) {
		var detail Actions
		tr.Find("tr").Each(func(col int, td *goquery.Selection) {
			table_length := td.Find("td, th").Length()
			td.Find("td").Each(func(col int, td *goquery.Selection) {

				if col == 0 && td.Text() != "" {
					if table_length == 3 {

						detail.ResourceType = strings.TrimSpace(td.Text())
					} else if table_length == 4 {
						detail.Description = strings.TrimSpace(td.Text())
					} else {

						detail.Action = strings.TrimSpace(td.Text())
					}

				} else if col == 1 && td.Text() != "" {
					if table_length == 3 {

						text := strings.TrimSpace(td.Text())
						text = re.ReplaceAllString(text, " ")
						text = strings.ReplaceAll(text, "\n", " ")
						detail.ConditionKeys = text
					} else {

						detail.Description = strings.TrimSpace(td.Text())
					}
				} else if col == 2 && td.Text() != "" {

					detail.Accesslevel = strings.TrimSpace(td.Text())

				} else if col == 3 {

					detail.ResourceType = strings.TrimSpace(td.Text())

				} else if col == 4 {

					text := strings.TrimSpace(td.Text())
					text = strings.ReplaceAll(text, "\n", ", ")
					detail.ConditionKeys = text

				}
			})

			if detail.Action != "" || detail.Description != "" || detail.ResourceType != "" || detail.ConditionKeys != "" || detail.Accesslevel != "" {
				details = append(details, detail)
			}
		})
	})
	return details, nil
}

// NOTE TO SELF: to pass sessions as parameter, you can use session neo4j.SessionWithContext
// and run query with session.Run(context.Background(), query, params)

func ingest_csv(ctx context.Context, driver neo4j.DriverWithContext, filename string, datatype string, fields []string) {
	query := "LOAD CSV FROM 'file:///" + filename + "' AS row WITH "

	row_names := ", "
	var rows []string

	for i := 0; i < len(fields); i++ {
		rows = append(rows, ("row[" + strconv.Itoa(i) + "] as " + fields[i]))
	}
	query += strings.Join(rows, row_names)

	query += (" MERGE (a:" + datatype + " {" + fields[0] + ":" + fields[0] + "})" +
		" ON CREATE SET ")

	for _, field := range fields {
		query += "a." + field + " = " + field + ", "
	}

	query += "a.layer = 0"

	neo4j.ExecuteQuery(ctx, driver, query,
		nil, neo4j.EagerResultTransformer,
		neo4j.ExecuteQueryWithDatabase("neo4j"))

	fmt.Println("[*] Added: ", filename)

}

func ingest_relationships(ctx context.Context, driver neo4j.DriverWithContext, file string, source_label string, source_field string, dest_label string, dest_field string, rel_name string) {

	query := "LOAD CSV FROM 'file:///" + file + "' AS row CALL WITH row " +
		"MERGE (s:" + source_label + " {{" + source_field + " +: row[0]}}) " +
		"ON CREATE SET s.inferred = true " +
		"MERGE (d:" + dest_label + " {{" + dest_field + " : row[1]}}) " +
		"ON CREATE SET d.inferred = true " +
		"MERGE (s) - [:" + rel_name + "  {{layer: 0}}] -> (d) " +
		"} IN TRANSACTIONS"

	neo4j.ExecuteQuery(ctx, driver, query,
		nil, neo4j.EagerResultTransformer,
		neo4j.ExecuteQueryWithDatabase("neo4j"))
	//executeQuery(context, driver, query, params, resultTransformer)
	/*
		error_check(err)

		fmt.Println("[+] Added: ", result.Summary)
	*/
}

func load_csvs_into_db(ctx context.Context, driver neo4j.DriverWithContext) {
	ingest_csv(ctx, driver, "awsmultivaluedprefix.csv", "AWSMultivalueOperator:UniqueName", []string{"name"})
	ingest_csv(ctx, driver, "awsresourcetypes.csv", "AWSResourceType:UniqueName", []string{"name", "arn", "regex"})
	ingest_csv(ctx, driver, "awsglobalconditionkeys.csv", "AWSConditionKey:UniqueName", []string{"name"})
	ingest_csv(ctx, driver, "awsconditionkeys.csv", "AWSConditionKey:UniqueName", []string{"name"})
	ingest_csv(ctx, driver, "awsactions.csv", "AWSAction:UniqueName", []string{"name", "access_level"})
	ingest_csv(ctx, driver, "awsservices.csv", "AWSService:UniqueName", []string{"name", "prefix", "url"})
	ingest_relationships(ctx, driver, "actions_to_resourcetypes_rels.csv", "AWSAction:UniqueName", "name", "ActsOn", "AWSResourceType:UniqueName", "name")

}

func create_constraint(ctx context.Context, driver neo4j.DriverWithContext, constraint_name string, label string, property string) {
	query := "CREATE CONSTRAINT " + constraint_name + " IF NOT EXISTS FOR (n: " + label + ") REQUIRE n." + property + " IS UNIQUE"

	neo4j.ExecuteQuery(ctx, driver, query,
		nil, neo4j.EagerResultTransformer,
		neo4j.ExecuteQueryWithDatabase("neo4j"))

	fmt.Println("[*] Created Constraint: " + constraint_name)
}

func create_relationship_constraint(ctx context.Context, driver neo4j.DriverWithContext, constraint_name string, rel_name string, unique_property_name string) {
	query := "CREATE CONSTRAINT " + constraint_name + " IF NOT EXISTS " +
		"FOR () - [r:" + rel_name + "] -() REQUIRE (r." + unique_property_name + " IS UNIQUE"

	neo4j.ExecuteQuery(ctx, driver, query,
		nil, neo4j.EagerResultTransformer,
		neo4j.ExecuteQueryWithDatabase("neo4j"))

}
func create_constraints_in_neo4j(ctx context.Context, driver neo4j.DriverWithContext) {

	create_constraint(ctx, driver, "awsactionconstraint", "AWSAction", "name")
	create_constraint(ctx, driver, "actionblobconstraint", "AWSActionBlob", "name")
	create_constraint(ctx, driver, "assumerolepolicyconstraint", "AWSAssumeRolePolicy", "hash")
	create_constraint(ctx, driver, "conditionconstraint", "AWSCondition", "hash")
	create_constraint(ctx, driver, "conditionvalueconstraint", "AWSConditionValue", "name")
	create_constraint(ctx, driver, "groupconstraint", "AWSGroup", "arn")
	create_constraint(ctx, driver, "inlinepolicyconstraint", "AWSInlinePolicy", "hash")
	create_constraint(ctx, driver, "managedpolicyconstraint", "AWSManagedPolicy", "arn")
	create_constraint(ctx, driver, "policydocumentconstraint", "AWSPolicyDocument", "hash")
	create_constraint(ctx, driver, "policyversionconstraint", "AWSPolicyVersion", "hash")
	create_constraint(ctx, driver, "roleconstraint", "AWSRole", "arn")
	create_constraint(ctx, driver, "statementconstraint", "AWSStatement", "hash")
	create_constraint(ctx, driver, "userconstraint", "AWSUser", "arn")
	create_constraint(ctx, driver, "resourceblobconstraint", "AWSResourceBlob", "name")
	create_constraint(ctx, driver, "tagconstraint", "AWSTag", "hash")
	create_constraint(ctx, driver, "uniquehashconstraint", "UniqueHash", "hash")

}
func create_indices(ctx context.Context, driver neo4j.DriverWithContext) {
	var query string

	query = "CREATE TEXT INDEX uniquehash IF NOT EXISTS " +
		"FOR (n:UniqueHash) ON (n.hash)"

	neo4j.ExecuteQuery(ctx, driver, query,
		nil, neo4j.EagerResultTransformer,
		neo4j.ExecuteQueryWithDatabase("neo4j"))

	query = "CREATE TEXT INDEX uniquearn IF NOT EXISTS " +
		"FOR (n:UniqueArn) ON (n.arn)"

	neo4j.ExecuteQuery(ctx, driver, query,
		nil, neo4j.EagerResultTransformer,
		neo4j.ExecuteQueryWithDatabase("neo4j"))

	query = "CREATE TEXT INDEX uniquename IF NOT EXISTS " +
		"FOR (n:UniqueName) ON (n.name)"

	neo4j.ExecuteQuery(ctx, driver, query,
		nil, neo4j.EagerResultTransformer,
		neo4j.ExecuteQueryWithDatabase("neo4j"))

	query = "CREATE TEXT INDEX statementeffect IF NOT EXISTS " +
		"FOR (s:AWSStatement) ON (s.effect)"

	neo4j.ExecuteQuery(ctx, driver, query,
		nil, neo4j.EagerResultTransformer,
		neo4j.ExecuteQueryWithDatabase("neo4j"))

	query = "CREATE TEXT INDEX actionname IF NOT EXISTS " +
		"FOR (s:AWSAction) ON (s.name)"

	neo4j.ExecuteQuery(ctx, driver, query,
		nil, neo4j.EagerResultTransformer,
		neo4j.ExecuteQueryWithDatabase("neo4j"))

	query = "CREATE TEXT INDEX actionblobname IF NOT EXISTS " +
		"FOR (s:AWSActionBlob) ON (s.name)"

	neo4j.ExecuteQuery(ctx, driver, query,
		nil, neo4j.EagerResultTransformer,
		neo4j.ExecuteQueryWithDatabase("neo4j"))

	query = "CREATE TEXT INDEX rolebname IF NOT EXISTS " +
		"FOR (s:AWSResource) ON (s.name)"

	neo4j.ExecuteQuery(ctx, driver, query,
		nil, neo4j.EagerResultTransformer,
		neo4j.ExecuteQueryWithDatabase("neo4j"))
}

func write_data_to_csv(filename string, data []string) {
	//output this all to a csv

	file, err := os.Create(filename)

	if err != nil {
		panic(err)
	}

	defer file.Close()
	writer := csv.NewWriter(file)
	defer writer.Flush()

	//Title of column for CSV
	header := []string{"name"}
	writer.Write(header)

	for _, value := range data {
		err := writer.Write([]string{value})
		if err != nil {
			panic(err)
		}
	}

}

func error_check(err error) {
	if err != nil {
		fmt.Println("Error with response:", err)
	}
}

func main() {
	/*
		//take in command line arguments
		var output_directory string
		var input_schema string

		flag.StringVar(&output_directory, "o", "/import/", "The directory to output csv files. This will default to import directory.")
		flag.StringVar(&output_directory, "output", "/import/", "The directory to output csv files. This will default to import directory.")

		flag.StringVar(&input_schema, "i", "", "The AWS input schema")
		flag.StringVar(&input_schema, "input-schema", "", "The AWS input schema")

		var (
			_, b, _, _ = runtime.Caller(0)
			basepath   = filepath.Dir(b)
		)
		output := (filepath.Join(basepath, "../../")) + output_directory


			//urls that will be used to grab services, policies, and actions
			service_url := "https://awspolicygen.s3.amazonaws.com/js/policies.js"
			base_url := "https://docs.aws.amazon.com/service-authorization/latest/reference/"
			service_json_url := base_url + "toc-contents.json"
	*/
	//context for neo4j database
	ctx := context.Background()
	//initialize database
	dbUri := "bolt://localhost:7687"
	driver, err := neo4j.NewDriverWithContext(
		dbUri,
		neo4j.NoAuth())
	error_check(err)

	defer driver.Close(ctx)

	// Create a session
	session := driver.NewSession(ctx, neo4j.SessionConfig{DatabaseName: "neo4j"})
	defer session.Close(ctx)
	/*
		services_metadata, err := get_service_metadata(service_url)
		error_check(err)

		actionMap := make(map[string]Actions)
		serviceMap := make(map[string]interface{})
	*/

	//awsglobalconditionkeys.csv
	//write_data_to_csv((output + "awsglobalconditionkeys.csv"), services_metadata.ConditionKeys)
	fmt.Println("[*] Condition keys written to awsglobalconditionkeys.csv")

	//awsoperators.csv
	//write_data_to_csv((output + "awsoperators.csv"), services_metadata.ConditionOperators)
	fmt.Println("[*] Operators written to awsoperators.csv")

	fmt.Println("[*] Gathering URLs of services")
	//href := get_services_json_href(service_json_url)

	fmt.Println("[*] Gathering Metadata of services")
	/*
		for title, url := range href {
			action_metadata, err := get_service_dict(base_url + url)
			error_check(err)
			actionMap = make(map[string]Actions)

			for _, items := range action_metadata {
				actionMap[items.Action] = items
			}
			serviceMap[title] = actionMap
			//fmt.Println(serviceMap[title])
		}
	*/
	//write json file
	fmt.Println("[*] Writing to json file")

	//convert servicemap to json
	//json_output, err := json.MarshalIndent(serviceMap, "", "  ")
	//error_check(err)

	/*
		//create and write to file
		aws_scheme_filewriter, err := os.Create("awsschema-test.json")
		error_check(err)

		defer aws_scheme_filewriter.Close()
		aws_scheme_filewriter.Write(json_output)
	*/
	fmt.Println("[*] Data written to awsschema.json")

	//initialize database
	fmt.Println("[*] Loading csv's into db")
	load_csvs_into_db(ctx, driver)

	fmt.Println("[*] Creating Constraints")
	create_constraints_in_neo4j(ctx, driver)

	fmt.Println("[*] Creating Indices")
	create_indices(ctx, driver)

	fmt.Println("[*] Complete")

}
